learning_observer/learning_observer/google.py

```diff
@@ -432,7 +432,7 @@ def extract_text_from_google_doc_json(
     length = j['body']['content'][-1]['endIndex']
     elements = [a.get('paragraph', {}).get('elements', []) for a in j['body']['content']]
     flat = sum(elements, [])
-    text_chunks = [f['textRun']['content'] for f in flat]
+    text_chunks = [f.get('textRun', {}).get('content', '') for f in flat]
```

Comment:
Here, I noticed that the text reconstruction was failing because some of the items in the array were style formatting and not actual edits/insertions so I only grabbed the text if they existed

Reply:
The code below it for align assumes the chunks exist so you will need to update that code as well to avoid errors.

Reply:
I'm not sure which specific piece of code you are referring to that needs additional error checking.
If you are referring to the code I posted below, the `text_chunks` will already be defined. If nothing is found, it will be an empty list.

Where is the issue?

```python
 if align:
        lengths = [f['endIndex'] - f['startIndex'] for f in flat]
        text_chunks = [_force_text_length(chunk, length) for chunk, length in zip(text_chunks, lengths)]
```

Reply:
Yes, that is correct. If `text_chunks` is empty, subsequent parts will be empty as well and the resultant `text` will be an empty string

Reply:
The issue is that you are iterating over flat again and not using the safe get.  As I found when debugging the prior code for the doc processor some of the entries did not have the text and wound up empty but still had spans.  As a consequence the two iterative passes did not match.  It is therefore cleaner to do a single iteration and call the force only when the text and spans are present.

Reply:
When running in cases where no text was present keeping the second iteration over flat created errors.  Here is the structure that does work:

```
    if align:
        for f in flat:
            text = f.get('textRun', {}).get('content', None)
            if text != None:
		length = f['endIndex'] - f['startIndex']
                text_chunks.append(_force_text_length(text, length))
    else:
        for f in flat:
            text = f.get('textRun', {}).get('content', None)
            if text != None:
                text_chunks.append(text)

```